rpc/jsonrpc/server/http_server.go

```diff
@@ -277,3 +280,49 @@ func Listen(addr string, maxOpenConnections int) (listener net.Listener, err err
 
 	return listener, nil
 }
+
+// Middleware
+
+// PreChecksHandler is a middleware function that checks the size of batch requests and returns an error
+// if it exceeds the maximum configured size. It also checks if the request body is not greater than the
+// configured maximum request body bytes limit.
+func PreChecksHandler(next http.Handler, config *Config) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// ensure that the current request body bytes is not greater than the configured maximum request body bytes
+		r.Body = http.MaxBytesReader(w, r.Body, config.MaxBodyBytes)
+
+		// if maxBatchSize is 0 then don't constraint the limit of requests per batch
+		// the default value is 0, and it cannot be negative because of the config validation
+		if config.MaxRequestBatchSize > 0 {
+			var requests []types.RPCRequest
+			var responses []types.RPCResponse
+			var err error
+
+			data, err := io.ReadAll(r.Body)
+			if err != nil {
+				res := types.RPCInvalidRequestError(nil, fmt.Errorf("error reading request body: %w", err))
+				_ = WriteRPCResponseHTTPError(w, http.StatusBadRequest, res)
+				return
+			}
+
+			err = json.Unmarshal(data, &requests)
+			// if no err it means multiple requests, check if the number of request exceeds
+			// the maximum batch size configured
+			if err == nil {
+				// if the number of requests in batch exceed the maximum configured then return an error
+				if len(requests) > config.MaxRequestBatchSize {
+					res := types.RPCInvalidRequestError(nil, fmt.Errorf("batch request exceeds maximum (%d) allowed number of requests", config.MaxRequestBatchSize))
+					responses = append(responses, res)
+					_ = WriteRPCResponseHTTP(w, responses...)
+					return
+				}
+			}
+
+			// ensure the request body can be read again by other handlers
+			r.Body = io.NopCloser(bytes.NewBuffer(data))
```

Comment:
shouldn't that be `r.Body = http.MaxBytesReader(w, r.Body, config.MaxBodyBytes)` otherwise we lose max body bytes restriction, no?

Reply:
I think the first line will catch that because it's outside the `if`, but I will do some test to confirm. Thanks for pointing this out.